## Introduction

### Node degree

Networks contain information about relationships between entities ("edges between nodes").
One key property of a node (especially in unweighted networks) is its degree, the number of connections a node makes with other nodes in the network.
Degree is a very important metric for differentiating between nodes and it is heavily relied-upon by a number of edge prediction (also known as "link prediction") features.
However, reliance on degree may not always be desirable.
Firstly, degree may be a source of bias in networks.
Secondly, an over-reliance on degree can lead edge prediction methods to make spurious or trivial predictions for some prediction tasks, failing to predict the insightful types of relationships desired.

Most biomedical data networks are imperfect representations of the true relationships.
Real networks often mistakenly include edges that do not exist and exclude edges that do exist.
How well a network represents the true relationships it attempts to represent depends on a number of factors, especially the methods used to generate the data in the network.
We define "degree bias" as the type of misrepresentation that occurs when the fraction of incorrectly existent/nonexistent relationships depends on degree.
Inspection bias indicates that entities are not uniformly studied [@doi:10.1038/nature04209], and it is likely to cause degree bias when networks are constructed using hypothesis-driven findings extracted from the literature, as newly-discovered relationships are not randomly sampled from the set of all true relationships.
The presence of inspection bias in a network means that well-studied entities have a higher fraction of their true relationships represented than less well-studied entities.
Depending on the type of data being represented, degree biases can also arise due to multifunctionality [@doi:10.1371/journal.pone.0017258], experimental methods, or other factors.
Figure {@fig:degree}A-C illustrates the problem of degree bias by showing node degree in separate networks for the same type of data.
Depending on the methods by which the represented data were generated, networks of the same type of data may assign very different degree to the same nodes, and they may have overall degree distributions that differ greatly.
The consequence of degree bias is that a difference in degree between two nodes may not reflect a difference in the number of true edges.
Reliance on degree may be unfavorable depending on the prediction/analysis task being conducted and the magnitude of possible degree bias in the data.

Another circumstance in which a reliance on degree may be unfavorable is when prediction of edges between lower-degree nodes is desired.
Many real networks have degree distributions that follow a power law, meaning they have a wide distribution of node degrees and none of their nodes could be considered representative of the remaining nodes [@doi:10.1242/jcs.02714].
Other networks may even have multimodal degree distributions or vastly different distributions for source and target degrees.
Figure {@fig:degree}B shows examples of degree imbalances for six biomedical networks.
High degree imbalance can lead especially high-degree nodes to dominate in predictions.
The result of such imbalance is that degree-based predictions are likely nonspecific, meaning the same set of predictions would appear to perform well for different prediction tasks.
As an analogy, Crow et al. [@doi:10.1073/pnas.1802973116] found that a single optimized gene ranking could be, "used to guess the hit list of any DE [differential expression] study without knowing anything about it".
Similarly, network edge prediction could become nonspecific when the top predicted edges for any node connect to the highest degree nodes in the network.
Depending on the prediction task and the desired types of predictions, edge predictions between very high degree nodes may be uninsightful and undesired.
Model evaluation can also be complicated by such predictions, as trivial predictions need not be incorrect, even if  they are not the desired outputs of a predictive model.
Rather, degree-based features must be included in the interpretation of predictions in order to disentangle desired from non-desired predictions and to more effectively evaluate and compare predictive models.

![**A.** Degree distributions of networks with and without degree bias can be very different.
  Protein-protein interaction (PPI) literature-derived network from STRING (Szklarczyk et al.) [@doi:10.1093/nar/gky1131], high-throughput-derived networks from Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
  Transcription factor-target gene (TF-TG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].
  BioRxiv data from Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] had edges split between those existing before and during/after 2018.
  **B.** Node degree in a literature-derived network is not linearly correlated with degree in a systematically-derived network.
  Assuming that the inclusion of an edge in the systematic network is independent of node degree, the systematic degree should be linearly correlated with the true number of node relationships.
  Edges found in literature-derived networks represent a non-random sample from systematically-derived edges.
  Shown are the same PPI and TF-TG networks as in subfigure A.
  **C.** Uniform random sampling edges produces node degrees that are linearly correlated with those in the unsampled network.
  Shown are the same PPI and TF-TG networks as in subfigure A.
  70% of literature edges were sampled with uniform probability.
  **D.** Degree distribution can vary greatly between networks, even between the source and target degree distributions within the same network.
  Shown are six networks from Hetionet [@doi:10.7554/eLife.26726].
](https://github.com/greenelab/xswap-analysis/raw/ff5c45d1c42b52156eb939ab67ac69b082176a88/img/fig1.degree_bias.png){#fig:degree width="100%"}


### Edge prediction

### Feature-degree correlation


## Methods

### Network permutation

### XSwap algorithm

Hanhijärvi, et al. presented XSwap [@doi:10.1137/1.9781611972795.67], an algorithm for the randomization ("permutation") of unweighted networks (Figure {@fig:algo}A).
The algorithm picks two existing edges at random, and if the edges constitute a valid swap, exchanges the targets between the edges (Figure {@fig:xswap}).

![Graphical representation of the XSwap algorithm applied to two edges.
The algorithm preserves both the source- and target-degree for all nodes.](images/xswap_figure.png){#fig:xswap width="50%"}

To allow greater flexibility, we modified the algorithm by adding two parameters, "`allow_self_loops`", and "`allow_antiparallel`" that allow a greater variety of network types to be permuted (Figure {@fig:algo}B).
Specifically, two chosen edges constitute a valid swap if they preserve degree for all four involved nodes and do not violate the above condition options.
The motivation for these generalizations is to make the permutation method applicable both to directed and undirected graphs, as well as to networks with different types of nodes, variously called multipartite, heterogeneous, or multimodal networks.
We provide documentation for parameter choices depending on the type of network being permuted in the GitHub repository (https://github.com/hetio/xswap).
The original algorithm and our proposed modification are given in Figure {@fig:algo}.

![
  **A.** XSwap algorithm due to Hanhijärvi, et al. [@doi:10.1137/1.9781611972795.67].
  **B.** Proposed modification to XSwap algorithm](images/xswap_algos.png){#fig:algo width="100%"}

### Edge prior

We introduce the "edge prior" to quantify the probability that two nodes are connected based only on their degree.
The edge prior can be estimated using the maximum likelihood estimate for the binomial distribution success probability--the fraction of permuted networks in which a given edge exists.
Based only on permuted networks, the edge prior does not contain any information about the true edges in the (unpermuted) network.
The edge prior is a numerical feature that can be computed for each node pair in a network, and we compared its ability to predict edges in three tasks, discussed below.

### Edge prior approximation

We also considered the possibility that the probability of an edge existing across permuted networks could be written as a closed form equation involving the node pair's degree.
We were unable to find a closed-form solution giving the edge prior without assuming independent node pairs, which we believe is incorrect for XSwap.
Nonetheless, we discovered a good approximation to the edge prior for networks with many nodes and relatively low edge density.

Let $m$ be the total number of edge in the network, and $d(u_i)$, $d(v_j)$ be the source and target degrees of a node pair, respectively.
A good approximation of the edge prior is given by the following:
\begin{equation}
    P_{i,j} = \frac{d(u_i) d(v_j)}{\sqrt{(d(u_i) d(v_j))^2 + (m - d(u_i) - d(v_j) + 1)^2}}
\end{equation}

Further discussion of this approximate edge prior and an derivation are available in [the supplement](#approx-prior-supp).

### Prediction tasks

### Degree-grouping

### Implementation and source code


## Results

![Discrimination figure](){#fig:discrimination width="100%"}

![Calibration figure](){#fig:calibration width="100%"}

## Discussion


## Conclusion
