## Introduction

### Node degree bias

Biomedical data networks are rarely, if ever, perfectly faithful representations of all true relationships.
False positives and false negatives are present in almost every type of biomedical network.
A number of factors can contribute to the number of connections that an entity makes in a certain network (the node's "degree").
Ideally, the main determinant of degree would be the number of true relationships.
Technical biases can also influence node degree, though these depend on the type of biomedical data being considered and may not always introduce degree bias.
Another major influence is the extent to which an entity has been studied relative to other entities in the network.
Inspection bias is the phenomenon in which entities are not uniformly studied [@doi:10.1038/nature04209], meaning that well-studied entities likely have a higher fraction of their true relationships represented in a network than less well-studied entities.
This particular kind of bias is likely when networks are constructed using hypothesis-driven findings extracted from the literature, as newly-discovered relationships are not randomly sampled from the set of all true relationships.

Bias in the degree structure of networks can be demonstrated by comparing node degree between literature-derived and unbiased networks in domains where unbiased experimentation methods exist.
Figure {@fig:degree-bias}A shows examples of degree distributions for different networks in the same domain.
Each of the three types of data shows a comparison between the degree distributions of two networks.
For PPI and TF-TG data, one of the two networks shown is produced via unbiased experimentation, and the other is based on relationships found in the literature.
Meanwhile, the bioRxiv co-authorship network was divided by whether relationships existing before 2018 or after.
While the co-authorship networks are both faithful representations of all true relationships for their time periods--meaning they do not exhibit a degree bias--the pairs of networks for both PPI and TF-TG data exhibit noticeable differences in degree distribution.
The differences in degree distribution arising from data generation methods are examples of degree bias in real data.

Figure {@fig:degree-bias}B-C, shows bias in individual node degree for the PPI and TF-TG networks.
On the one hand, when edges are uniformly sampled there remains a linear correlation between degree in the networks (Figure {@fig:degree-bias}B).
On the other hand, non-random sampling--as low-throughput, hypothesis-driven scientific experimentation represents in this example--can introduce degree bias in both directions relative to unbiased investigations.
Figure {@fig:degree-bias}C shows an example of degree biases between networks of the same type of data.

![**A.** Degree distributions of networks with and without degree bias can be very different.
  Protein-protein interaction (PPI) literature-derived network from STRING (Szklarczyk et al.) [@doi:10.1093/nar/gky1131], high-throughput-derived networks from Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
  Transcription factor-target gene (TF-TG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].
  BioRxiv data from Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] had edges split between those existing before and during/after 2018.
  **B.** Uniform random sampling edges produces node degrees that are linearly correlated with those in the unsampled network.
  70% of literature edges were sampled with uniform probability.
  **C.** Node degree in a literature-derived network is not linearly correlated with degree in a systematically-derived network.
  Assuming that the inclusion of an edge in the systematic network is independent of node degree, the systematic degree should be linearly correlated with the true number of node relationships.
  Edges found in literature-derived networks represent a non-random sample from systematically-derived edges.](https://github.com/zietzm/xswap-analysis/raw/08ff7f4803ba940cca38d31def4b20f006ccec36/img/fig1.degree_bias.png){#fig:degree-bias width="100%"}

Many domains do not have unbiased experimental methods, either because such methods have not yet been developed or because such methods are not possible.
For example, compound-disease treatment relationships in human patients can not be elucidated via large scale, unbiased experimentation due to ethical standards.
Without unbiased data, currently available networks have an unquantifiable node degree bias.
The potential biases in node degree motivate the development and use of methods to quantify and correct for the effects of degree in network analysis, inference, and link prediction.

### Unsupervised link prediction

We focus on (biomedical) network link prediction.
In particular, we are interested in predicting new edges with specificity, such that the predictions reflect more than simply node degree--a potential source of bias.
Unbiased prediction is challenging because the target degree distribution (the distribution of degree in true relationships) cannot, in general, be predicted from a network with degree bias.
For example, Figure {@fig:degree-bias}A shows cases in which the bias in literature-derived networks is toward either higher or lower degree.
A simple adjustment, therefore, is not possible.

Our focus is on unsupervised methods for prioritizing new links, because supervised methods require labeled training data, which is not available in domains without comprehensive, unbiased, experimental methods.
Network reconstruction tasks allow supervised methods to be trained to reconstruct the network.
However, because the target degree distribution cannot be known, reconstruction tasks train methods to predict links from a different degree distribution than the target.
Therefore, training a supervised method via network reconstruction only trains the method to predict sampled edges, which are not, in general, reflective of the edges missing between biased and unbiased networks.
To enable prediction without degree bias, a method is needed which can account for degree effects without knowing a priori the direction or magnitude of the bias.

### Feature-degree correlation

Network analysis methods can examine networks at many scales, including node, node-pair, community, and whole-network.
In this investigation, we restrict our attention to node-pair features used for link prediction, though the principles and permutation framework are also applicable to other types of network analysis.

Bias in a network's degree distribution can be toward either higher or lower degree (Figures {@fig:degree-bias}A,{@fig:degree-bias}C).
To make unbiased predictions, therefore, features used for prediction should not be strongly influenced by degree, either positively or negatively.
In practice, though, many features commonly used for link prediction are correlated with degree (Figure {@fig:features-degree}).

![Common link-prediction features are correlated with node degree.
Shown are five features (Supplemental table 2) computed on literature-derived protein-protein interaction network [@doi:10.1093/nar/gky1131] and the geometric mean of the node pair's source and target degrees.
All five features show a positive relationship with degree, though the magnitude of this correlation is highly variable.](https://github.com/greenelab/xswap-analysis/raw/08ff7f4803ba940cca38d31def4b20f006ccec36/img/fig2.feature_degree.png){#fig:features-degree width="100%"}

Some link prediction features are functions of basic node pair properties like degree.
In such cases, the effects of node degree can be predicted or derived directly.
For example, the resource allocation index is the sum of inverse degree of common neighbors between source and target nodes (in the symmetric case), while preferential attachment is the product of source and target degree [@doi:10.1140/epjb/e2009-00335-8; @doi:10.1145/1065385.1065415].
In many other cases, the effects of degree on link prediction methods cannot be determined so easily.
Therefore, it is important to have a more general method for comparing the effects of node degree on link prediction methods and the prediction tasks themselves.


## Methods

### Network permutation

Network permutation eliminates true connections in a network, but can be customized to preserve certain aspects of a network.
We propose the use of degree-preserving network permutation to eliminate true connections and higher-order connectivity information while retaining node degree.
Permuted networks can be used to evaluate the fraction of link prediction performance that is attributable entirely to degree.
A method which has strong reconstruction performance on a true network and very poor performance on permuted networks likely relies on higher-order information that is eliminated in permutation.
Conversely, a link-prediction method that shows identical predictive performance on true and permuted networks likely relies on information that is preserved during permutation.
The extent to which a link-prediction method relies on degree can be determined by comparing it on a true network and degree-preserving permutations of that network.
Network permutation is flexible and can be broadly applied because it generates networks that are identical in format to a real network.
This means that any network-based method could be analyzed for reliance on degree by using permuted networks.

### XSwap algorithm

Hanhij√§rvi, et al. presented XSwap [@doi:10.1137/1.9781611972795.67], an algorithm for the randomization ("permutation") of unweighted networks (Figure {@fig:algo}A).
The algorithm picks two existing edges at random, and if the edges constitute a valid swap, exchanges the targets between the edges (Figure {@fig:xswap}).

![Graphical representation of the XSwap algorithm applied to two edges.
The algorithm preserves both the source- and target-degree for all nodes.](images/xswap_figure.png){#fig:xswap width="50%"}

To allow greater flexibility, we modified the algorithm by adding two parameters, "`allow_self_loops`", and "`allow_antiparallel`" that allow a greater variety of network types to be permuted (Figure {@fig:algo}B).
Specifically, two chosen edges constitute a valid swap if they preserve degree for all four involved nodes and do not violate the above condition options.
The motivations for these generalizations are to make the permutation method applicable both to directed and undirected graphs, as well as to networks with different types of nodes, variously called multipartite, heterogeneous, or multimodal networks.
The original algorithm and our proposed modification are given in Figure {@fig:algo}.

![
  **A.** XSwap algorithm due to Hanhij√§rvi, et al. [@doi:10.1137/1.9781611972795.67].
  **B.** Proposed modification to XSwap algorithm](images/3_xswap_algorithm_pseudocode/algos-merged.png){#fig:algo width="100%"}

### Edge prior

We introduce the edge prior as a feature quantifying the affinity for two nodes to share an edge in permuted networks.
The edge prior can be estimated using the maximum likelihood estimate for the binomial distribution success probability--the fraction of permuted networks in which a given edge exists.
Based only on permuted networks, the edge prior does not contain any information about the true edges in the (unpermuted) network.
The edge prior is a numerical feature that can be computed for each node pair in a network, and we compared its ability to predict edges in three tasks, discussed below.

We also considered the possibility that the probability of an edge existing across permuted networks could be written as a closed form equation involving the node pair's degree.
Without the assumption of independent node pairs that we believe is incorrect for XSwap, we were unable to find a closed-form solution giving the edge prior, because node pairs cannot be treated as independent.
Nonetheless, we discovered a good approximation to the edge prior for networks with many nodes and relatively low edge density.

Let $m$ be the total number of edge in the network, and $d(u_i)$, $d(v_j)$ be the source and target degrees of a node pair, respectively.
A good approximation of the edge prior is given by the following:
\begin{equation}
    P_{i,j} = \frac{d(u_i) d(v_j)}{\sqrt{(d(u_i) d(v_j))^2 + (m - d(u_i) - d(v_j) + 1)^2}}
\end{equation}

Further discussion of this approximate edge prior and an derivation are available in the supplement.

### Prediction tasks

We performed three prediction tasks to assess the performance of the edge prior, and we compared two additional features, a scaled product of source and target degree (scaled to the range [0, 1]), and our approximation of the edge prior.
For the first two prediction tasks, we used 20 biomedical networks from the Hetionet heterogeneous network [@doi:10.7554/eLife.26726.001] that had at minimum 2000 edges.
In the first task, we computed the degree-based prediction features (edge prior, scaled degree product, and prior approximation), and predicted the original edges in the network.
Negative samples in the first task were node pairs in which an edge was not present in the original network.
To assess the methods' predictive performances, we computed the area under the receiver operating characteristic curve for all three features.
In the second task, we sampled 70% of edges from each of the networks, computed features on the sampled network, then attempted to predict held-out edges.
For this task, negative samples were node pairs in which an edge did not exist in either original or sampled network, while positive samples were only those node pairs without an edge in the sampled network but with an edge in the original network.

The third task evaluated the ability of the edge prior to generalize to new degree distributions.
We used two domains--protein-protein interactions (PPI) and transcription factor-target gene (TF-TG) relationships--where networks are available that were generated through different methods: literature curation of low-throughput, hypothesis-driven research and high-throughput, systematic, hypothesis-free experimentation.
For the PPI networks, we used the STRING network, which incorporates literature-mining to find relationships [@doi:10.1093/nar/gky1131] and a combination of the high-throughput, proteome-scale interaction networks due to Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
We used a transcription factor-target gene (TF-TG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and a high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].
As a further basis of comparison, we added a time-resolved co-authorship network which we partitioned by time to create two separate networks.
The co-authorship network represented bioRxiv preprints, and it was created using the Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] database that was generated by crawling the bioRxiv server.
Unlike the other two, the co-authorship network does not have degree bias, as the network faithfully represents all true co-author relationships.
We include this network to offer a comparative prediction task in which the degree distributions between training (papers before 2018) and testing (posted to bioRxiv during or after 2018) do not differ (Figure {@fig:degree-bias}A).
For each of the three data sources--PPI, TF-TG, co-authorship--we had two networks, one in which we expect degree bias and one in which we do not.
The goal of the third prediction task is to determine feature generalizability for network reconstruction between different degree distributions, especially predicting a network without degree bias using features from a degree-biased network.
Further information about the networks used can be found in Supplemental table 1.

### Degree-grouping

Our method for degree-preserving permutation switches almost all edges in the network.
Because only degree is preserved, nodes with equal degree can be grouped when summarizing features.
For example, in link prediction methods, two node pairs with the same source and target degree have link prediction feature values in permuted networks that are drawn from the same distribution.
Each node pair's feature values in permuted networks can be augmented in this way to have more permuted feature values than permuted networks.
For nodes with a common degree, degree grouping can greatly increase the effective number of permutations, even though only a smaller number of permuted networks were generated.
We utilized degree grouping throughout our analysis, including in computing the edge prior.

### Implementation and source code

We implemented the modified XSwap algorithm as a Python package, with the actual edge swap mechanism implemented in C++ for greater speed.
In addition to functions that permute networks (represented as edge lists), the package contains utilities for computing the edge prior, converting a network between adjacency matrix and edge list formats, and for assigning unique identifiers to nodes.
The code for our method of degree-preserving network permutation has been made freely available ([https://github.com/hetio/xswap](https://github.com/hetio/xswap)), as has the code for the analysis, figure generation ([https://github.com/greenelab/xswap-analysis](https://github.com/greenelab/xswap-analysis)), and manuscript ([https://github.com/greenelab/xswap-manuscript](https://github.com/greenelab/xswap-manuscript)).


## Results

In the first prediction task, we computed three features--the XSwap edge prior, an approximation to the edge prior, and the (scaled) product of source and target node degree--on networks from Hetionet.
We used the features in attempting to reconstruct the 20 networks themselves.
The XSwap-derived edge prior is able to reconstruct many of the networks with a high level of performance, as measured by the AUROC.
Of the 20 individual networks we extracted from Hetionet, 17 had an edge prior self-reconstruction AUROC >= 0.95, with the highest reconstruction AUROC at 0.9971.
Meanwhile, the lowest self-reconstruction performance (AUROC = 0.7697) occurred in the network having the fewest node pairs.

![
  **A.** AUROC of network reconstruction by prediction task
  **B.** Sampling network edges generally reduces performance of reconstructing the unsampled network.](https://github.com/zietzm/xswap-analysis/raw/08ff7f4803ba940cca38d31def4b20f006ccec36/img/fig3.combined_auroc.png){#fig:all_auroc width="85%"}

The three features used in the comparison are highly rank correlated (median > 0.99999 across metaedges).
Because AUROC takes only rank into account, the three features have very similar AUROC reconstruction performance values for the first, second, and third prediction tasks (max difference < 0.027).
Nonetheless, we found that the edge prior was the superior feature in 12 of 20 networks, indicating marginally superior discriminatory performance relative to the other features.
While the discriminatory abilities are similar, the features are very different in their levels of calibration.
We find that the XSwap-derived edge prior is very well calibration for all networks in the first and second tasks, and it shows the best calibration of the three features for each of the prediction tasks (Figure {@fig:cal}A).
As the edge prior does not have any information regarding the true edges in the network, these results indicate that degree sequence alone is highly informative and can allow the reconstruction of edges, and they validate the use of the edge prior for this purpose.

![
  **A.** Calibration curves for full network reconstruction. 20 networks shown are from Hetionet.
  **B.** Calibration curves for sampled network reconstruction. 20 networks shown are from Hetionet.
  **C.** Boxplots summarizing individual Hetionet network calibrations. Calibration metric is from the two-component decomposition of the Brier score, and a lower score indicates superior calibration.](https://github.com/zietzm/xswap-analysis/raw/08ff7f4803ba940cca38d31def4b20f006ccec36/img/fig4.calibration.png){#fig:cal width="100%"}

In the second task, we sampled 70% of edges from each network, computed features using the sampled networks, then predicted the edges (or lack thereof) in the unsampled network for node pairs without an edge in the sampled network.
Since the holdout edges were sampled randomly without replacement from the set of all edges, the degree distribution for the holdout edges is essentially the same as the original network's degree distribution (for example, Figure {@fig:degree-bias}B).
Unlike in the first task, however, edges that were present in the sampled network are not tested and therefore do not show up in the performance metrics.

As in the first task, the results from the second prediction task demonstrate a high level of performance for degree-sequence-based node pair features (Figure {@fig:all_auroc}B).
Reconstruction performance was expectedly lower in the second task than in the first, though many networks were still able to be reconstructed with a high level of performance.
In the second task, the edge prior computed on the sampled network was able to reconstruct the unsampled network with an AUROC of greater than 0.9 in 14 of 20 networks.
We find that the three features have roughly similar calibrations in the second task as they had in the first.
Surprisingly, though, scaled degree improved its average calibration across networks for the second task relative to the first.
For all tasks, the edge prior again showed the best calibration relative to the other features.
In fact, in the self-reconstruction task the edge prior had calibration an order of magnitude superior to the other two features.
Similarly to the features in the first task, node pair features computed on sampled networks in the second prediction task were also highly rank-correlated.

We find that there is an association between the ability of the edge prior to reconstruct a network and the Gini coefficient of the degree among all nodes in the network.
This correlation was found in full-network self-reconstruction, but the association appears to disappear for reconstruction of sampled networks.
This result indicates a potential association between node degree inequality and the predictiveness of degree for specific edges, though a greater understanding of this phenomenon is required.

In the third prediction task, we compute the three edge prediction features for paired networks representing data from PPI, TF-TG, and bioRxiv bioinformatics pre-print co-authorship.
The goal of the task was to compare predictive performance across different degree distributions for the same type of data.
We find that the task of predicting systematically-derived edges using a network with degree bias is significantly more challenging than network reconstruction.
We find consistently lower performance in predicting systematically-derived edges than reconstructing the original network (Figure {@fig:all_auroc}A).
The edge prior was not able to predict the separate PPI network at all, with an AUROC of roughly 0.5.
Only slightly better was its performance in predicting the separate TF-TG network, at an AUROC of 0.59.
We find superior performance in predicting the co-authorship relationships (AUROC 0.75), which was expected as the network being predicted shared roughly the same degree distribution as the network on which the edge prior was computed.
The results of the third prediction task show that a difference in degree distribution between the network on which features are computed and the network to be predicted can make prediction significantly more challenging.

## Discussion

In this work we concern ourselves with the task of link prediction and the effects of node degree.
Network node degree can strongly influence network analysis and link prediction metrics, and node degree can be subject to large biases.
To quantify the influence of node degree on network reconstruction and link prediction more generally, we devised a network permutation framework using a modified version of the XSwap algorithm.
Our method quantifies the link prediction performance attributable to the network's degree sequence alone.
Depending on the type of data contained, a network's degree sequence may or may not rank-order entities according to their actual relative number of connections.
As in a previously discussed example, if a network reflects low-throughput, hypothesis-driven experimentation toward a particular goal (for example, specific genes of interest), then a difference in degree between two entities in the network may not reflect a difference in the number of true connections.
Differences in degree between entities in a network with degree bias do not necessarily imply that the entities have a different number of true relationships.
Our method seeks to quantify the effect of degree on link prediction tasks and methods.

To quantify the effect of degree in a particular link prediction task, we propose a feature called the "edge prior".
The edge prior is computed using degree-preserving network permutation, and its value is attributable entirely to the network's degree sequence.
Three link prediction tasks demonstrated that the edge prior can predict edges in many networks, sometimes with extremely high discrimination performance.
We also found that the edge prior performs similarly to--albeit slightly better than--two degree-based features we used for comparison.
In all tasks, we found the edge prior to be well-calibrated and to show superior calibration relative to the simpler features.
The edge prior's dependence on degree alone and strong predictive performance in terms both of discrimination and calibration lead us to believe that the edge prior represents a suitable baseline for network link prediction methods when network degree may be subject to bias.

Broadly, we propose the use of network permutation to quantify the effect of degree in link prediction.
However, there are a number of possible ways to quantify the effect of degree for a prediction task, including two other features to which we compared the edge prior.
In our comparison, we found the edge prior to be superior to the other methods, in calibration and--to a lesser extent--in discrimination.
A more important advantage of permutation over the other methods is its ability to interrogate other link prediction methods themselves.
While other degree-based features can only quantify the effect of degree on the prediction task, our network permutation method allows the effect of degree to be quantified both for the prediction task and for other prediction methods.
By generating complete networks through permutation, our method allows any link prediction method to be applied to a true network and permutations of the network.
At the level of individual node pairs, the value of a link prediction feature in the true network can be compared to the distribution of the specific feature across permuted networks.
In this way, the effect of degree on a link prediction method can be determined as well.

For a link prediction task to be affected by degree means that degree-based features perform strongly on the task.
Moreover, networks that can be reconstructed with high performance using the edge prior have many nodes, and nodes with a large number of different degrees.
In our comparison of 20 networks from Hetionet, for example, all networks with a mean number of source and target nodes over 2,500 had self-reconstruction AUROC performance >= 0.95.
The interpretation of this finding is that larger networks potentially can be better predicted through degree alone compared to smaller networks.
This increasing baseline level of performance can be discovered by finding the performance of the edge prior for a supervised link prediction task.
However, we also found that attempting to predict edges from a different degree distribution as the one on which features were computed significantly increases the difficulty of the task.
Tasks like the PPI prediction we attempted underline the large influence degree can have on prediction tasks and validate our permutation method as a way to determine when and how much prediction tasks are affected by degree.

The method we present is applicable to many types of networks, generalizing easily to bipartite, multipartite (or "heterogeneous"), or directed networks.
Permutations are especially useful for unsupervised edge prediction and node pair prioritization in heterogeneous networks.
To permute a heterogeneous network, each edge type can be permuted separately as if it were an independent bipartite network.
Such a procedure would ensure that degree is preserved for each node on a node-type basis as well.
In heterogeneous networks, features can be computed on a path-type (or "metapath") basis, meaning that unrelated  degree biases could be included in every feature.


## Conclusion

Degree bias is manifest in many networks of biomedical data.
Despite this, many link prediction methods are highly correlated with degree.
We propose a network permutation framework to quantify the effects of degree toward link prediction tasks and link prediction methods.
We quantify degree's performance toward a specific task using a permutation-based feature we term the "edge prior".
Based only degree alone, the edge prior represents a useful baseline performance level for tasks where data may be subject to degree bias.
We perform three link prediction tasks that demonstrate the edge prior to be a strong predictive method, in terms both of discrimination and calibration.
Additionally, our method can quantify the effect of degree on other link prediction methods.
Our results highlight the strong effect that degree has on link prediction, visualizes examples of the biases to which degree can be subject, and proposes a permutation-based framework for quantifying the effect of node degree.
