## Introduction

### Node degree bias

Biomedical data networks are rarely, if ever, perfectly faithful representations of all true relationships.
False positives and false negatives are present in almost every type of biomedical network.
A number of factors can contribute to the number of connections that an entity makes in a certain network (the node's "degree").
Ideally, the main determinant of degree would be the number of true relationships.
Technical biases can also influence node degree, though these depend on the type of biomedical data being considered and may not always depend on degree.
Another major influence is the extent to which an entity has been studied relative to other entities in the network.
Inspection bias is the phenomenon in which entities are not uniformly studied [@doi:10.1038/nature04209], meaning that well-studied entities likely have a higher fraction of their true relationships represented in a network than less well-studied entities.
This particular kind of bias is likely when networks are constructed using hypothesis-driven findings extracted from the literature, as newly-discovered relationships are not randomly sampled from the set of all true relationships.

Bias in the degree structure of networks can be demonstrated by comparing node degree between literature-derived and unbiased networks, in domains where unbiased experimentation methods exist.
Figure {@fig:degree-bias}B shows examples of degree distributions for different networks in the same domain.
Three domains are represented, and each has two networks whose degree distributions are compared.
For PPI and TF-TG data, each relationship type has two networks--one produced via unbiased experimentation and one based on relationships found in the literature.
Meanwhile, the bioRxiv co-authorship network was divided by whether relationships existing before 2018 or after.
While the co-authorship networks are both faithful representations of all true relationships for their time periods--meaning they do not exhibit a degree bias--the pairs of networks for both PPI and TF-TG data exhibit noticeable differences in degree distribution.

Figure {@fig:degree-bias}B-C, shows bias in individual node degree for the PPI and TF-TG networks.
On the one hand, when edges are uniformly sampled there remains a linear correlation between degree in the networks (Figure {@fig:degree-bias}B).
On the other hand, non-random sampling--as is done in low-throughput, hypothesis-driven scientific experimentation--can introduce bias in both directions relative to systematic investigations.
Figure {@fig:degree-bias}C shows an example of degree biases that can be thought-of as degree-biased sampling.

![**A.** Degree distributions of biased and unbiased networks can be very different.
  Protein-protein interaction (PPI) literature-derived network from STRING (Szklarczyk et al.) [@doi:10.1093/nar/gky1131], high-throughput-derived networks from Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
  Transcription factor-target gene (TF-TG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].
  BioRxiv data from Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] split between edges existing before and during/after 2018.
  **B.** Uniform random sampling edges produces sampled degrees that are linearly correlated.
  70% of literature edges were sampled with uniform probability.
  **C.** Node degree in a literature-derived network is not linearly correlated with degree in a systematically-derived network.
  Assuming that the inclusion of an edge in the systematic network is independent of node degree, the systematic degree should be linearly correlated with the true number of node relationships.
  Edges found in literature-derived networks represent a non-random sample from systematically-derived edges.](images/combined.png){#fig:degree-bias width="100%"}

Many domains do not have unbiased experimental methods, either because such methods have not yet been developed or because such methods are not possible.
For example, compound-disease treatment relationships in human patients can not be elucidated via large scale, unbiased experimentation due to ethical standards.
Without unbiased data, currently available networks have an unquantifiable node degree bias.
It is therefore helpful to use methods that can quantify and correct for the effects of degree on the methods used for network analysis, inference, and prediction.

### Unsupervised link prediction

We focus on (biomedical) network link prediction.
In particular, we are interested in predicting new edges with specificity, such that the predictions reflect more than simply node degree--a potential source of bias.
Unbiased prediction is challenging because the target degree distribution (the distribution of degree in true relationships) cannot, in general, be predicted from a network with degree bias.
For example, Figure {@fig:degree-bias}A shows cases in which the bias in literature-derived networks is toward either higher or lower degree.

Our focus is on unsupervised methods for prioritizing new links, because supervised methods require labeled training data, which is not available in domains without comprehensive, unbiased, experimental methods.
Network reconstruction tasks allow supervised methods to be trained to reconstruct the network.
However, because the target degree distribution cannot be known, reconstruction tasks train methods to predict links from a different degree distribution than the target.
Therefore, training a supervised method via network reconstruction only trains the method to predict sampled edges, which are not, in general, reflective of the edges missing between biased and unbiased networks.
To enable prediction without degree bias, a method is needed which can account for degree bias without knowing a priori the direction or magnitude of the bias.

### Feature-degree correlation

Network analysis methods are varied and can examine networks at many scales, including node, node-pair, community, and whole-network.
In this investigation, we restrict our attention to node-pair features used for purposes of link prediction, though the principles and permutation framework are also applicable to other types of network analyses.

Bias in a network's degree distribution can be toward either higher or lower degree (Figures {@fig:degree-bias}A,{@fig:degree-bias}C).
To make unbiased predictions, therefore, features used for prediction should not be strongly influenced by degree, either positively or negatively.
In practice, though, many features commonly used for link prediction are correlated with degree (Figure {@fig:features-degree}).

![Common link-prediction features are correlated with node degree.
Shown are five features (Supplemental table 2) computed on literature-derived protein-protein interaction network [@doi:10.1093/nar/gky1131] and the geometric mean of the node pair's source and target degrees.
All five features show a positive relationship with degree, though the magnitude of this correlation is highly variable.
For example, preferential attachment and random walk with restart (RWR) appear to have stronger relationships with degree than the Jaccard index.](https://github.com/greenelab/xswap-analysis/raw/bd40eb244ffe37454dc08099f5742ef3e962c70d/img/2_link_prediction_feature_degree_correlation/without_prior.png){#fig:features-degree width="100%"}

Some link prediction features are functions of basic node pair properties like degree.
These features (Supplemental table 2) can allow the effect of node degree to be predicted or derived directly.
For example, the resource allocation index is the sum of inverse degree of common neighbors between source and target nodes (in the symmetric case), while preferential attachment is the product of source and target degree [@doi:10.1140/epjb/e2009-00335-8; @doi:10.1145/1065385.1065415].
Many features and link-prediction methods, however, cannot be interrogated to determine, _a priori_, the effect that node degree will have.
Therefore, it is important to have a more general method for comparing the relative effects of node degree on features and other link-prediction methods.


## Methods

### Network permutation

Network permutation enables the generation of networks that share some properties of a network of interest while eliminating other information.
We propose the use of degree-preserving network permutation to eliminate true connections and higher-order connectivity information while retaining node degree and overall network degree sequence.
Permuted networks--generated via degree-preserving network permutation--can be used to evaluate the performance of link prediction methods that is attributable entirely to node degree and overall degree sequence.
That is, a method which has strong performance on a true network and very poor performance on permuted networks likely utilizes higher-order information that is eliminated in permutation.
Conversely, a link-prediction method that shows identical predictive performance on true and permuted networks likely relies on information that is preserved during permutation.
The extent to which a link-prediction method relies on degree can be determined by comparing it on a true network and permutations of that network.
Network permutation can be broadly applied because it generates networks that are identical in format to a real network.
This means that any network-based method could be analyzed for reliance on degree by using permuted networks.

### XSwap algorithm

Hanhijärvi, et al. presented XSwap [@doi:10.1137/1.9781611972795.67], an algorithm for the randomization ("permutation") of unweighted networks (Figure {@fig:algo}A).
The algorithm picks two existing edges at random, and if the edges constitute a valid swap, exchanges the targets between the edges (Figure {@fig:xswap}).

![Graphical representation of the XSwap algorithm applied to two edges.
The algorithm preserves both the source- and target-degree for all nodes.](images/xswap_figure.png){#fig:xswap width="50%"}

To allow greater flexibility, we modified the algorithm by adding two parameters, "`allow_self_loops`", and "`allow_antiparallel`" that allow a greater variety of network types to be permuted (Figure {@fig:algo}B).
Specifically, two chosen edges constitute a valid swap if they preserve degree for all four involved nodes and do not violate the above condition options.
The motivations for these generalizations are to make the permutation method applicable to both directed and undirected graphs, as well as between networks with different types of nodes, variously called multipartite, heterogeneous, or multimodal networks.
The original algorithm and our proposed modification are given in Figure {@fig:algo}.

![**A.** XSwap algorithm due to Hanhijärvi, et al. [@doi:10.1137/1.9781611972795.67]. **B.** Proposed modification to XSwap algorithm](images/3_xswap_algorithm_pseudocode/algos-merged.png){#fig:algo width="100%"}

### Edge prior

We introduce the edge prior as a feature quantifying the affinity for two nodes to share an edge in permuted networks.
The edge prior can be estimated using the maximum likelihood estimate for the binomial distribution success probability--the fraction of permuted networks in which a given edge exists.
Based only on permuted networks, the edge prior does not contain any information about the true edges in the (unpermuted) network.
The edge prior is a numerical feature that can be computed for each node pair in a network, and we compared its ability to predict edges in three tasks: reconstruction of the original network, reconstruction of a holdout set of edges that were removed before permutation, and reconstruction of a separate network of the same type of data as the original, but with a potentially different degree distribution.
With these three tasks, we are able to evaluate well degree predicts the original network, edges from the same degree distribution, and edges from a different degree distribution but with the same nodes.

### Edge prior approximation

Degree preserving network randomization using the XSwap algorithm preserves only the network's degree sequence.
Therefore, we considered the possibility that the probability of an edge existing across permuted networks could be written as a closed form equation involving the network's adjacency matrix.
We simplified the problem by assuming that node pairs can be treated as independent.
This simplification does not correctly model the true behavior of the algorithm, but it is a good estimate in networks with ...

**Explanation/evaluation of when the approximation is good and when it is not. Also, add a quick derivation of approximation here as well.**

### Prediction tasks

We performed three prediction tasks to assess the performance of the edge prior.
Because the edge prior is based only on the network degree sequence, we compared its performance to two additional features, a scaled product of source and target degree, and the closed-form approximation of the edge prior.
For the first two prediction tasks, we used sub-networks from the Hetionet heterogeneous network that had at minimum 2000 edges.
In the first task, we computed the degree-based prediction features (edge prior, scaled degree product, and prior approximation), and predicted the original edges in the network.
Negative samples in the first task were node pairs in which an edge was not present in the original network.
To assess the methods' predictive performances, we computed the area under the receiver operating characteristic curve for all three features.
In the second task, we sampled 70% of edges from each of the networks, computed features on the sampled network, then attempted to predict held-out edges.
For this task, negative samples were node pairs in which an edge did not exist in either original or sampled network, while positive samples were only those node pairs without an edge in the sampled network but with an edge in the original network.

The third task evaluated the ability of the edge prior to generalize to new degree distributions.
In particular, we used two domains--protein-protein interactions (PPI) and transcription factor-target gene (TF-TG) relationships--with fundamentally different methods of generating networks: literature curation of low-throughput, hypothesis-driven research and high-throughput, systematic, hypothesis-free experimentation.
For the PPI networks, we used the STRING network, which incorporates literature-mining to find relationships [@doi:10.1093/nar/gky1131] and a combination of the high-throughput, proteome-scale interaction networks due to Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
We used a transcription factor-target gene (TF-TG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and a high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].
As a further basis of comparison, we added a time-resolved co-authorship network which we partitioned by time to create two separate networks.
The co-authorship network represented bioRxiv preprints, and it was created using the Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] database that was generated by crawling the bioRxiv server.
Unlike the other two, the co-authorship network does not have degree bias, as the network faithfully represents all true co-author relationships.
We include this network to offer a comparative prediction task in which the degree distributions between training (papers before 2018) and testing (posted to bioRxiv during or after 2018) do not differ (Figure {@fig:node-degree-bias}).

To summarize the third task, we used three different types of biomedical information networks for comparison purposes-- a protein-protein interaction (PPI) network, a transcription factor-target gene (TF-TG) network, and a co-authorship network for papers on bioRxiv in the bioinformatics section.
For each of the three data sources--PPI, TF-TG, co-authorship--we had three networks, representing a degree-biased (or previous) network, a sampled version (in which a fraction of edges are dropped), and a systematic (or future) network, in which we do not expect degree bias.
Information about the number of nodes and edges in these networks is available in Supplemental table 1.

### Degree-grouping

Our method for degree-preserving permutation switches almost all edges in the network.
Because only degree is preserved, nodes with equal degree can be grouped when summarizing features.
For example, in link prediction methods, two node pairs with the same source and target degree have link prediction feature values in permuted networks that are drawn from the same distribution.
Each node pair's feature values in permuted networks can be augmented in this way to have more permuted feature values than permuted networks.
For nodes with a common degree, degree grouping can greatly increase the effective number of permutations, even though only a smaller number of permuted networks were generated.

### Implementation and source code

We implemented the modified XSwap algorithm as a Python package, with the actual edge swap mechanism implemented in C++ for greater speed.
In addition to functions that permute networks (represented as edge lists), the package contains utilities for computing the edge prior, converting a network between adjacency matrix and edge list formats, and for assigning unique identifiers to nodes.
The code for our method of degree-preserving network permutation has been made freely available ([https://github.com/hetio/xswap](https://github.com/hetio/xswap)), as has the code for the analysis, figure generation ([https://github.com/greenelab/xswap-analysis](https://github.com/greenelab/xswap-analysis)), and manuscript ([https://github.com/greenelab/xswap-manuscript](https://github.com/greenelab/xswap-manuscript)).


## Results

The XSwap-derived edge prior feature is able to reconstruct many of the networks with a high level of performance, as measured by the AUROC.
Of the 20 individual networks we extracted from Hetionet, 17 had an edge prior self-reconstruction AUROC of over 0.9, with the highest reconstruction AUROC at 0.9971.
Meanwhile, the lowest self-reconstruction performance (AUROC = 0.7697) occurred in the network having the fewest node pairs.
As the edge prior does not have any information regarding the true edges in the network, these results indicate that degree sequence alone is highly informative and can allow the reconstruction of edges.

In the first prediction task, we computed three features--the XSwap edge prior, an approximation to the edge prior, and the (scaled) product of source and target node degree--on networks from Hetionet.
We used the features in attempting to reconstruct the 20 networks themselves.
The three features used are highly rank correlated (median > 0.99999 across metaedges).
Because AUROC takes only rank into account, the three features have very similar AUROC reconstruction performance values for the first, second, and third prediction tasks (max difference < 0.027).
(**Maybe footnote: "Subtle but pervasive XSwap is better**)
While the discriminatory abilities are very similar, the features are very different in their levels of calibration.
We find that the XSwap-derived edge prior is very well calibration for all networks in the first and second tasks, and it shows the best calibration of the three features for each of the prediction tasks (Figure {@fig:cal}A).

![**A.** Calibration curves for full network reconstruction. 20 networks shown are from Hetionet. **B.** Calibration curves for sampled network reconstruction. 20 networks shown are from Hetionet. **C.** Mean network calibration by network type for three features **D.** Boxplots summarizing individual Hetionet network calibrations](images/calibration_combined.png){#fig:cal width="100%"}

In the second task, we sampled 70% of edges from each network, computed features using the sampled networks, then predicted the edges (or lack thereof) in the unsampled network for node pairs without an edge in the sampled network.
Since the holdout edges were sampled randomly without replacement from the set of all edges, the degree distribution for the holdout edges is essentially the same as the original network's degree distribution (**?Figure?**).
Unlike in the first task, however, edges that were present in the sampled network are not tested and therefore do not show up in the performance metrics.

As in the first task, the results from the second prediction task demonstrate a high level of performance for degree-sequence-based node pair features (Figure {@fig:auroc}).
Reconstruction performance was expectedly lower in the second task than in the first, though many networks were still able to be reconstructed with a high level of performance.
In the second task, the edge prior computed on the sampled network was able to reconstruct the unsampled network with an AUROC of greater than 0.9 in 14 of 20 networks.
We find that the three features have similar calibrations in the second task as they had in the first, though, surprisingly, scaled degree improved its average calibration across networks for the second task relative to the first.
For all tasks, the edge prior showed the best calibration relative to the other features.
In fact, in the self-reconstruction task the edge prior had calibration an order of magnitude superior to the other two features.
Similarly to the features in the first task, node pair features computed on sampled networks in the second prediction task were also highly rank-correlated.

![AUROC between full and sampled networks](images/full_vs_sampled_auroc.png){#fig:auroc width="50%"}

We find that there is an association between the ability of the edge prior to reconstruct a network and the Gini coefficient of the degree among all nodes in the network.
This correlation was found in full-network self-reconstruction, but the association appears to disappear for reconstruction of sampled networks.
This result indicates a potential association between node degree inequality and the predictiveness of degree for specific edges, though a greater understanding of this phenomenon is required.

In the third prediction task, we compute the three edge prediction features for paired networks representing data from PPI, TF-TG, and bioRxiv bioinformatics pre-print co-authorship.
For the PPI and TF-TG tasks, the paired networks do not share their degree distributions, as in both cases one network represents edges from low-throughput, hypothesis driven research and the other represents edges from unbiased, systematic, hypothesis-free experiments.
The co-authorship network was split between relationships that existed as of before 2018, those that exist as of present, and a sampled version of the relationships that existed as of before 2018.
Unlike the two experimental networks, the three co-authorship networks all have a similar degree distribution (Supplemental Figure {@fig:supp-dists}).
We include the comparison to illustrate a plausible situation in which the target network for link prediction has a similar degree distribution to the network available for computing features.

![AUROC between prediction tasks](images/auroc_by_task.png){#fig:auroc_by_task width="50%"}

We find that the task of predicting systematically-derived edges using a network with degree bias is more challenging than network reconstruction.
That is, we find consistently lower performance in predicting systematically-derived edges than reconstructing the original network (Figure {@fig:auroc_by_task}).

**Bit more explanation needed here. Wrap up the results section.**

## Discussion

In this work we concern ourselves with the task of link prediction and the effects of node degree.


Previous work has shown that networks of biomedical data are biased toward well-studied entities [@doi:10.1038/nature04209].
No networks of biomedical data contain all the true relationships.
Within the networks that exist and exhibit the bias previously reported, well-studied entities have a disproportionate fraction of their true relationships represented in the network.
Meanwhile, poorly-characterized entities have a smaller fraction of their relationships represented in the network.
This can lead to a situation in which two entities have the same number of true relationships but have a large difference in the number of connections in the network, due to one of the two being better-studied than the other.
Unbiased methods seek to identify equally edges connected to well- and poorly-characterized nodes.

Entities can have a relatively lower or higher number of connections in literature-derived networks when compared to systematic evaluations.
We have two potential explanations for these relationships.
First, where the degree in a literature-derived network is higher (PPI in Figure {@fig:features-degree}), one can conclude that the experimental method has a lower recall than small-scale experimentation.
Previous work supports this conclusion for the protein-protein interaction network used as an example in the present work.
Huang and Bader [@doi:10.1093/bioinformatics/btn640] found that the Y2H method is subject to a high false negative rate.
Moreover, this finding makes sense as low-throughput experimentation can offer a much higher resolution of an individual PPI, especially in an _in vivo_, tissue-specific, or otherwise context-dependent setting that cannot be captured by the Y2H.
In the second case, entities have lower degree in literature-derived networks.
In general, this indicates that low-throughput experimentation has, thus far, identified only a small fraction of the true relationships.

In either case, the difference in degree distributions between networks derived from low-throughput and systematic evaluations is reflective of a bias in the degree to which entities have been studied and the effectiveness of experimental methods.
Differences in degree between entities in a network with degree bias do not necessarily imply that the entities have a different number of true relationships.
Moreover, since the features used for many applications--specifically link prediction--are influenced by degree, it is helpful to consider the counterfactual for prediction where systematic evaluation methods are unfeasible or do not exist.


We also introduce the edge prior, a permutation-derived node-pair feature that shows promise for link-prediction.
The edge prior is the fraction of permutations in which a particular edge appears.
Because it is based only on degree-preserving permutations of a network, the edge prior is attributable solely to the degree structure of a network.
We find that the edge prior is a predictive feature for certain tasks, performing better than a feature specificially-designed for predicting directed edges in the transcription factor-target gene network (Figure {@fig:ROC}).
This result validates the fact that degree is highly predictive for some tasks, despite the biases to which it may be subject.

While the finding that the edge prior is helpful for many tasks validates our network permutation method as a helpful method for analysis, a limitation of our method remains in that, for some of the prediction tasks, neither features nor corrected features were highly predictive.
Moreover, for supervised link prediction applications, we do not expect corrected features to be, in general, superior or sufficiently superior to warrant their use.
The utility of degree-preserving randomization and corrected features is in unsupervised network prediction and analysis tasks where the data available have a degree bias.
Moreover, networks in any domain where relationship data are not generated through unbiased means will be subject, to some extent, to degree bias.

The method we present is applicable to many types of networks, generalizing easily to bipartite, multipartite (or "heterogeneous"), or directed networks.
Permutations are especially useful for unsupervised edge prediction and node pair prioritization in heterogeneous networks.
To permute a heterogeneous network, each edge type can be permuted separately as if it were an independent bipartite network.
Such a procedure would ensure that degree is preserved for each node on a node-type basis as well.
In heterogeneous networks, features can be computed on a path-type (or "metapath") basis, meaning that unrelated  degree biases could be included in every feature.
Because an entire network can be permuted, our method allows the generation of a counterfactual distribution for heterogeneous networks in precisely the same fashion as for simple networks.
Finally, though features are computed for node pairs in the example of link prediction we present, the empirical significance method can be generalized to node, community, or network-level features as well.


## Conclusion

Inspection bias is manifest in many networks of biomedical data.
Well-studied entities can have many more relationships than poorly-studied entities that may have an equal number of relationships.
Moreover, while degree can be highly biased in networks, many link prediction methods are highly correlated with degree.
This motivates a method to account and correct for degree's effect on features and predictions.
We propose the use of degree-preserving network permutation to estimate feature values (on a node pair level) attributable entirely to degree.
Additionally, we propose an empirical significance scoring method to evaluate the likelihood that a feature value could arise entirely due to degree.
Adjusting for degree in the way we propose generates "corrected" features that are no longer linearly-correlated with degree but are as- or more effective in prioritizing systematically-derived edges.
Finally, we introduce a broadly-applicable, permutation-based link prediction feature that we term the, "edge prior".
This feature shows excellent performance for graph reconstruction and outperforms a specifically-created feature in predicting edges in a directed network.
Our results highlight the strong effect that degree has on link prediction, visualizes examples of the biases to which degree can be subject, and proposes a permutation-based framework for quantifying and correcting for the effect of node degree.
