## Background

### Node degree bias

Networks of biomedical data often have node degrees that are biased with respect to the extent to which certain entities have been studied.
Compared to a hypothetical network of the true relationships, biomedical data networks have edges that are missing in a non-uniform way, where some nodes have a much higher fraction of their true edges while others have almost none of their true edges.
This particular kind of bias can occur when networks are constructed using data extracted from the literature.
As the extent to which a biomedical entity has been studied does not necessarily imply its number of true connections, less well-studied entities will have degrees that are biased toward being significantly lower than their true number of connections, when compared to better-studied entities.
This phenomenon of edges missing in a non-uniform manner is due to "inspection bias", in which certain entities or classes of entities are studied in greater detail than others [@doi:10.1038/nature04209].

Bias in the structure of networks can be demonstrated in domains where unbiased experimentation methods exist.
While systematic methods are not free of all bias, the bias introduced should be independent of degree.
If the biases of high-throughput experimentation are roughly degree-independent, then the degree distributions of networks produced via comprehensive, systematic experimentation resemble the degree distributions of true relationships.
When such unbiased data are available, the degree bias in literature-derived networks can be quantified by comparing degree between literature-derived and unbiased networks.

Figure {@fig:node-degree-bias} gives examples of degree distribution bias in five networks, two protein-protein interaction networks (one of which is literature-derived, and the other of which was constructed via a high-throughput screen of all human proteins), two transcription factor-target gene networks (one of which is literature-derived, and the other of which was constructed from a ChIP-seq screen), and a co-authorship network based on bioinformatics preprints posted on bioRxiv. Figure {@fig:sampling_vs_systematic} shows bias in individual node degree for the PPI and TFTG networks. On the one hand, when edges are uniformly sampled there remains a linear correlation between degree in the networks. On the other hand, non-random sampling--as is done in low-throughput scientific experimentation--can introduce bias in both directions relative to systematic investigations.

![Degree distributions of biased and unbiased networks can be very different. Protein-protein interaction (PPI) literature-derived network from STRING (Szklarczyk et al.) [@doi:10.1093/nar/gky1131], high-throughput-derived networks from Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050]. Transcription factor-target gene (TFTG) literature-derived network from Han et al. [@doi:10.1093/nar/gkx1013] and high-throughput network from Lachmann et al. [@doi:10.1093/bioinformatics/btq466].  BioRxiv data from Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421].](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/1_degree_bias/histogram.png){#fig:node-degree-bias width="100%}

![Degree by sampling method. In the left column, 70% of literature edges were sampled with uniform probability. In the right column, literature derived degrees are over or under the corresponding systematically-derived degree.](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/7_well_studied_degree_bias/hexbin.png){#fig:sampling_vs_systematic width="85%"}

Networks created through degree-unbiased experimentation have node degrees that are linearly correlated with the number of true relationships.
Many domains, however, do not have unbiased experimental methods, either because such methods have not yet been developed or because such methods are not possible, in principle.
For example, compound-disease treatment relationships in human patients can not be elucidated via large scale, unbiased experimentation.
Without unbiased data, the available networks on which analysis methods can be applied (the data themselves) have node degree bias of an unquantifyable magnitude.
It is therefore helpful to use methods that can quantify and correct for the effects of degree on the methods used for network analysis, inference, and prediction.

### Feature-degree correlation

Network analysis methods are varied and can examine networks at varying scales, including node, node-pair, community, and whole-network.
In this investigation, we restrict our attention to node-pair features used for purposes of link prediction, though the principles are also applicable to other types of network analyses.

Bias in a network's degree distribution can be toward either higher or lower degree (Figure {@fig:node-degree-bias}).
To make unbiased predictions, therefore, features used for prediction should not be strongly influenced by degree, either positively or negatively.
In practice, though, many features commonly used for link prediction are correlated with degree (Figure {@fig:features-degree}).

![Common link-prediction features are correlated with node degree. Shown are five features computed on literature-derived protein-protein interaction network [@doi:10.1093/nar/gky1131] and the geometric mean of the node pair's source and target degrees. All five features show a positive relationship with degree, though the magnitude of this correlation is highly variable. For example, preferential attachment and random walk with restart (RWR) appear to have stronger relationships with degree than the Jaccard index.](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/2_link_prediction_feature_degree_correlation/without_prior.png){#fig:features-degree width="100%"}

Some link prediction features are functions of basic node pair properties like degree.
These features can allow the effect of node degree to be predicted or derived directly.
For example, the resource allocation index is the sum of inverse degree of common neighbors between source and target nodes (in the symmetric case), while preferential attachment is the product of source and target degree [@doi:10.1140/epjb/e2009-00335-8; @doi:10.1145/1065385.1065415].
Many features and link-prediction methods, however, cannot be interrogated to determine, _a priori_, the effect that node degree will have.
Therefore, it is important to have a more general method for comparing the relative effects of node degree on features and other link-prediction methods.

## Methods

### Network permutation

<!-- Need a discussion about what counterfactual would look like.
Wouldn't expect zero features in "null condition".
Should explain what the "null condition" should be, logically, and why
permutations capture that.

Also can't compare raw features between entities-->


The raw values of features computed for a given network entity depend both on relevant and non-relevant information, including extraneous and biased structural properties of the network.
While features may depend heavily on extraneous information, unsupervised applications seeking to order entities should rely as much as possible on the relevant information contained in the network.
To eliminate the bias due to network structure, we propose a comparison to the counterfactual condition in which a feature's value depends only on the potentially biasing network degree sequence.
This comparison allows a feature's true value to be given a corresponding "significance" score, which indicates how likely the feature would occur by chance given only degree sequence.
We term the empirical feature significance score the "corrected" feature, as it has been corrected for the contribution of degree to the raw feature.

Because relative degree is not, in general, reflective of the relative true degree of an entity, and because degree is highly influential in many features used for link prediction, we believe that degree is a confounding variable in biomedical link prediction.
To eliminate these confounding effects, we propose comparing entities not by their raw feature values, but by features corrected for confounding structural information through the use of meaningful counterfactual distributions. 
To quantify the effects of degree on network-based methods, we propose a comparison between a network of actual data and degree-preserving permutations of the network.
By comparing a network-based feature-extraction method between the true network and its permutations, our method isolates the contribution of network degree sequence and thereby quantifies the effect of degree on predictive/analysis methods.

The specifics of this comparison depend on the specific predictive or inference task being performed.
We use link prediction as an example application that illustrates several of the ways in which network permutation can be useful.
For the purposes of demonstration, we will focus our attention on two link prediction features: RWR, and the Jaccard index.

### XSwap algorithm

Hanhijärvi, et al. presented XSwap [@doi:10.1137/1.9781611972795.67], an algorithm for the randomization ("permutation") of unweighted networks (Figure {@fig:algo}A).
The algorithm picks two existing edges at random, and if the edges constitute a valid swap, exchanges the targets between the edges (Figure {@fig:xswap}).
To allow greater flexibility for heterogeneous biomedical networks with various types of relationships, we added two option parameters, "`allow_self_loops`", and "`allow_antiparallel`" that allow a greater range of possibly valid swaps (Figure {@fig:algo}B).
Specifically, two chosen edges constitute a valid swap if they preserve degree for all four involved nodes and do not violate the above condition options.
The motivations for these generalizations are to make the permutation method applicable to both directed and undirected graphs, as well as between networks with different types of nodes, variously called multipartite, heterogeneous, or multimodal networks.
The original algorithm and our proposed modification are shown in Figure {@fig:algo}.

![**A.** XSwap algorithm due to Hanhijärvi, et al. [@doi:10.1137/1.9781611972795.67]. **B.** Proposed modification to XSwap algorithm](images/3_xswap_algorithm_pseudocode/algos-merged.png){#fig:algo width="100%"}

![XSwap algorithm applied to two edges. The algorithm preserves both the source- and target-degree for all nodes, making it generalizable to directed, undirected, and bipartite networks.](images/xswap_figure.png){#fig:xswap width="50%"}

### Unsupervised prediction task

To evaluate the effectiveness of network  permutation for unsupervised link prediction, we performed two comparisons: a network reconstruction task and the prediction of unbiased relationships using a biased network.
To create the sampled network, we randomly dropped 30% of the network's edges.
Because each edge is equally likely to have been dropped, this method does not introduce new degree bias, and the resulting network has a degree distribution that is essentially equivalent to the source network `Supplemental figure`.
To predict the edges in a network without degree bias, we computed features on literature-curated networks that are subject to degree bias.
In each comparison, features were computed for all node pairs, but edge prediction was only carried out for node pairs that were not self-loops and for which an edge did not already exist in the network on which features were computed.

`Explain three feature categories. Feature, no edge info, corrected feature.`

### Data used for comparisons

We used three different types of biomedical information networks for comparison purposes-- a protein-protein interaction (PPI) network, a transcription factor-target gene (TFTG) network, and a co-authorship network for papers on bioRxiv in the bioinformatics section.
For each of the three data sources--PPI, TFTG, co-authorship--we had three networks, representing a degree-biased (or previous) network, a sampled version (in which a fraction of edges are dropped), and a systematic (or future) network, in which we do not expect degree bias.
For the PPI networks, we used the (`?mostly literature-curated`) STRING network [@doi:10.1093/nar/gky1131] and a combination of the high-throughput, proteome-scale interaction networks due to Rual et al. [@doi:10.1038/nature04209] and Rolland et al. [@doi:10.1016/j.cell.2014.10.050].
The bioRxiv co-authorship network was created using the Rxivist [@doi:10.1101/515643; @doi:10.5281/zenodo.2566421] database that was generated by crawling the bioRxiv server.
Unlike the other two, the co-authorship network does not have the equivalent of "literature-derived" and "high-throughput".
We include this network to offer a comparative prediction task in which the degree distributions between training (papers before 2018) and testing (posted to bioRxiv during or after 2018) do not differ (Figure {@fig:node-degree-bias}).


### Features

We used three previously-developed features for the unsupervised edge prediction task: random walk with restart (RWR), the Jaccard index, and an inference score developed for link prediction in directed graphs [@raw:inference]. RWR was implemented as defined in [@raw:rwr], with the steady-state distribution ($\mathbf{x}_i$) given as follows, where $c$ is the probability of a restart, $A$ is the normalized adjacency matrix, and $\mathbf{y}_i$ is a vector with zero at all entries except $i$, where it is equal to one:

\begin{equation}
\mathbf{x}_i = (1-c) \mathbf{A} \mathbf{x}_i + c \mathbf{y}_i
\end{equation}

We utilized the graph Laplacian method [@raw:laplacian] to normalize the adjacency matrix in a way that ensures results are symmetric, meaning $x_{i,j} = x_{j,i}$.

The Jaccard index is the fraction of the intersection and union of neighbors for two nodes. Let $k(u)$ be the set of neighbor nodes for node $u$. Then the Jaccard index for nodes $u$ and $v$ is

\begin{equation}
J(u, v) = \frac{|k(u) \cap k(v)|}{|k(u) \cup k(v)|}
\end{equation}

The inference score used for link prediction on the directed graphs was applied to the TFTG network.
To compute this score, let $A(u)$ denote the set of vertices that node $u$ points to and $D(u)$ the set of nodes that point to $u$.

\begin{equation}
S_{INF}(u, v) = \frac{|A(u) \cap D(v)|}{|A(u)|} + \frac{|D(u) \cap D(v)|}{|D(u)|}
\end{equation}

These features were chosen because they are not exact functions of degree, and the effect of degree on these features is not obvious.
RWR and the Jaccard index are appropriate for link prediction in simple graphs like the bioRxiv co-authorship network and PPI, but they are not appropriate for prediction in directed or bipartite graphs like the TF-TG network.
Their limitation in this sense is due to the fact that no nodes in separate classes of a bipartite graph share any neighbors, and RWR does not have a stationary distribution when walks must traverse two classes of nodes.

We introduce a further link prediction feature, termed the "edge prior", that is computed via network permutation.
This feature is estimated by the fraction of degree-preserving permutations in which a given edge exists.
The flexibility of our modified XSwap algorithm makes the edge prior computable on most types of networks, with the exception of multigraphs.
The edge prior's predictive performance is also evaluated for the unsupervised prediction tasks.

### Implementation and source code

We implemented the modified XSwap algorithm as a Python library, with the actual edge swap mechanism implemented in C++ for greater speed.
The code for our method of degree-preserving network permutation has been made freely available ([https://github.com/greenelab/xswap](https://github.com/greenelab/xswap)), as has the code for the analysis, figure generation ([https://github.com/greenelab/xswap-analysis](https://github.com/greenelab/xswap-analysis)), and manuscript ([https://github.com/greenelab/xswap-manuscript](https://github.com/greenelab/xswap-manuscript)).

## Results

### Comparison between networks

We computed link-prediction features for all node pairs in three types of biomedical networks--PPI, TFTG, and bioinformatics preprint co-authorship.
For each of the PPI and TFTG network types there were three networks on which features were computed, a degree-biased network, a sampled version of the degree-biased network, in which a fraction of edges were dropped, and a systematic network, in which no degree bias is expected.
The co-authorship network was split between relationships that existed as of before 2018, those that exist as of present, and a sampled version of the relationships that existed as of before 2018.
Unlike the two experimental networks, the three co-authorship networks all have a similar degree distribution `Supplemental figure`.
We include the comparison to illustrate a plausible situation in which the target network for link prediction has a similar degree distribution to the network available for computing features.

We computed RWR and the Jaccard index on the PPI and co-authorship networks and an inference score developed for link prediction in directed graphs [@raw:inference] on the TFTG network.
We used our network permutation algorithm to generate 1000 degree-preserving permutations of each of these networks, and we computed the same set of features on each permutation of each of the nine individual networks (three for each of PPI, TFTG, co-authorship).
By comparing features computed on the actual networks to the features computed on degree-preserving permutations of the networks, we were able to compute "corrected" features, which adjust the original features relative to the distribution of each node pair feature if only the network's degree sequence is retained.
Specifically, we defined the permutation-corrected node pair features as the fraction of permuted networks in which the node pair's feature value is greater than or equal to its value in the true network.

The features computed across true and permuted networks were evaluated for their ability to predict edges in the test networks corresponding to each data type.
In particular, for each network type (PPI, TFTG, co-authorship), we evaluated the ability of each sub-network's features to predict new edges in other networks.
The train/test network pairings evaluated were as follows: sampled/biased, biased/systematic, and systematic/biased.
Since the goal of our method is not to maximize reconstruction performance but ability to generalize to networks with different degree distributions, we are particularly interested in feature performance for the biased/systematic prediction task.
The results of these comparisons are shown in Figure {@fig:ROC}.

![Predictive performance toward two tasks: reconstruction of edges removed with uniform probability and prediction of systematic edges using a degree-biased network derived from the literature. ](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/4_ROC/reconstruction_systematic.png){#fig:ROC width="100%"}

We find that the task of predicting systematically-derived edges given a network with degree bias is more challenging than network reconstruction.
Nonetheless, while raw features show the greatest predictive performance for reconstruction in the bioRxiv and PPI networks (measured in terms of the ROC curve), we find that corrected features perform at least as well as raw features for the prediction of systematic edges.

The inference score does not show strong performance for predicting directed edges in the TFTG network, either in the reconstruction or systematic prediction tasks.
We were surprised to find that the edge prior shows superior performance for both tasks.
Additionally, we were surprised to find that the mean value of the inference score for each node pair in permuted networks shows superior performance to the inference score itself.
This result is unexpected because permutations do not contain true information, meaning that the mean inference score across permutations is attributable solely to degree.

`more explanation here`

We also found in some cases that permuted networks alone were superior to the true networks for link prediction.
In particular, we found that two metrics were often highly informative for link prediction, even though they contain no information about the real relationships in the network.
We term the simplest permutation-based feature the, "edge prior".
This feature is the probability of a particular edge existing in permuted networks with a given degree sequence, and it is estimated by the fraction of permuted networks in which the edge existed.
Additionally, for each node pair we computed the mean feature values across permuted networks.
This feature represents the expected feature value for a node pair in random networks with the given degree sequence.


![Comparison of degree-feature relationship between raw link prediction features and features corrected for degree using counterfactual distributions computed on permuted networks](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/5_corrected_features_not_correlated/points.png){#fig:feature-vs-corrected width="75%"}

<!-- Above figure but histograms of degrees for top, say, 1% of the values (ie top features or lowest p-value) -->

### Performance of the algorithm

![Percent of edges swapped vs number of attempts at edge swaps. "Density" here means the number of edges divided the number of node pairs or "potential edges".](https://github.com/greenelab/xswap-analysis/raw/2e173fc5cd21f3958617c7d10ab5cae48dfad1c9/img/6_xswap_percent_swapped_iterations/lines_continuous.png){#fig:swap-percent width="85%"}

<!-- Make above plot but with asymptotic value relative to degree. Could be part B to above part A. -->

<!-- Network density vs (fractional) attempts to 50% (or some cutoff) changed -->
<!-- Is it possible to estimate the fraction of the random graph space that is being explored by XSwap? -->
<!-- Is there a closed-form solution for the probability of an edge after a certain number of swaps? -->
<!-- Other methods to get similar results? Chung-Lu, "configuration model"? etc. -->

## Discussion

In this work we concern ourselves with the task of link prediction and the effects of node degree.
We find that many common link prediction methods are correlated with degree (Figure {@fig:features-degree}).
This result is unsurprising, as several link prediction features are computed as functions of degree (`Supplement`).
The correlation between degree and link prediction features raises an issue, however, when one considers the potential bias in node degree.
Previous work has shown that networks of biomedical data are biased toward well-studied entities [@doi:10.1038/nature04209].
No networks of biomedical data contain all the true relationships.
Within the networks that exist and exhibit the bias previously reported, well-studied entities have a disproportionate fraction of their true relationships represented in the network.
Meanwhile, poorly-characterized entities have a smaller fraction of their relationships represented in the network.
This leads to a situation in which two entities could have the same number of true relationships but have a large difference in the number of connections in the network, due to one of the two being better-studied than the other.
Unbiased methods seek to identify equally edges among well- and poorly-characterized nodes.

Entities can have a lower or higher relative number of connections in literature-derived networks when compared to systematic evaluations.
We have two potential explanations for these relationships.
First, where the degree in a literature-derived network is higher (PPI in Figure {@fig:features-degree}), one can conclude that the experimental method has a lower recall than small-scale experimentation.
Previous work supports this conclusion for the protein-protein interaction network used as an example in the present work.
Huang and Bader [@doi:10.1093/bioinformatics/btn640] found that the Y2H method is subject to a high false negative rate.
This finding explains the relationship in the top right panel of Figure {@fig:sampling_vs_systematic}.
Moreover, this finding makes sense as low-throughput experimentation can offer a much higher resolution of an individual PPI, especially in an _in vivo_, tissue-specific, or otherwise context-dependent setting that cannot be captured by the Y2H.
In the second case, entities have lower degree in literature-derived networks.
In general, this indicates that low-throughput experimentation has, thus far, identified only a small fraction of the true relationships.

In either case, the difference in degree distributions between networks derived from low-throughput and systematic evaluations is reflective of a bias in the degree to which entities have been studied and the effectiveness of experimental methods.
Differences in degree between entities in a network with degree bias do not imply that the entities have a different number of true relationships.
Moreover, since the features used for many applications--specifically link prediction--are influenced by degree, it is helpful to consider the counterfactual for prediction where systematic evaluation methods are unfeasible or do not exist.

We propose the use of empirical significance scores to "correct" features for the contribution of extraneous or biasing structural information.
Our proposed score is the fraction of permutations on which a more extreme feature value is observed compared to the true network.
These scores can be interpreted as the probability that an observed feature will occur due exclusively to the degree sequence of the network.
Low values of this score indicate that the feature is much higher than would be expected due only to the network's degree structure.
With degree contribution removed, the corrected features are attributable to higher-order structural information like community structure, as all such higher-order structure is eliminated during network permutation.

We also introduce a new, permutation-derived node-pair feature that shows promise for link-prediction in some applications--the so-called "edge prior".
The edge prior is the fraction of permutations in which a particular edge appears.
Because it is based only on degree-preserving permutations of a network, the edge prior is attributable solely to the degree structure of a network.
We find that the edge prior is a predictive feature for certain tasks, performing better than a feature specificially-designed for predicting directed edges in the transcription factor-target gene network (Figure {@fig:ROC}).
This result validates the fact that degree is highly predictive for some tasks, despite the biases to which it may be subject.

While the finding that the edge prior is helpful for many tasks validates our network permutation method as a helpful method for analysis, a limitation of our method remains in that, for some of the prediction tasks, neither features nor corrected features were highly predictive.
Moreover, for supervised link prediction applications, we do not expect corrected features to be, in general, sufficiently superior to warrant their use.
The utility of degree-preserving randomization and corrected features is in unsupervised network prediction and analysis tasks where the data available have a degree bias.
`For example ... ?`
The method is applicable to many types of networks, generalizing easily to bipartite, multipartite (or "heterogeneous"), or directed networks.
In the example of link prediction we present, features are computed for node pairs.
The empirical significance method can be generalized to node, community, or network-level features as well.


## Conclusion

